File: scraper/dse_scraper.py
#!/usr/bin/env python3
"""
DSE (Dhaka Stock Exchange) Data Scraper

Automatically fetches stock data from DSE website and other sources.
No manual data entry required!
"""

import requests
import json
import pandas as pd
from datetime import datetime, timedelta
from typing import Dict, List, Any, Optional
import time
import os
from pathlib import Path


class DSEScraper:
    """Scrape real-time and historical data from DSE"""

    def __init__(self, cache_dir: str = "data/cache"):
        """Initialize scraper with caching"""
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(parents=True, exist_ok=True)

        # DSE endpoints
        self.base_url = "https://www.dsebd.org"

        # Alternative data sources
        self.yahoo_base = "https://query1.finance.yahoo.com/v7/finance"
        self.investing_base = "https://api.investing.com/api"

    def get_latest_price(self, ticker: str) -> Dict[str, Any]:
        """Get latest price data for a ticker"""

        # Method 1: Try DSE API endpoint
        try:
            url = f"{self.base_url}/php_graph/market_price.php"
            response = requests.get(url, timeout=10)

            if response.status_code == 200:
                data = response.text
                # Parse the response for the ticker
                return self._parse_dse_current(data, ticker)
        except Exception as e:
            print(f"DSE API error: {e}")

        # Method 2: Try alternative endpoint
        try:
            url = f"{self.base_url}/latest_share_price_scroll_l.php"
            response = requests.get(url, timeout=10)

            if response.status_code == 200:
                return self._parse_price_table(response.text, ticker)
        except Exception as e:
            print(f"Alternative endpoint error: {e}")

        return {}

    def get_historical_data(self, ticker: str, days: int = 365) -> List[Dict[str, Any]]:
        """Get historical OHLCV data"""

        # Check cache first
        cache_file = self.cache_dir / f"{ticker}_history.json"
        if cache_file.exists():
            cache_age = time.time() - cache_file.stat().st_mtime
            if cache_age < 3600:  # 1 hour cache
                with open(cache_file, 'r') as f:
                    return json.load(f)

        data = []

        # Method 1: DSE CSV endpoint
        try:
            end_date = datetime.now()
            start_date = end_date - timedelta(days=days)

            url = f"{self.base_url}/day_end_archive.php"
            params = {
                'startDate': start_date.strftime('%Y-%m-%d'),
                'endDate': end_date.strftime('%Y-%m-%d'),
                'inst': ticker,
                'archive': 'data'
            }

            response = requests.get(url, params=params, timeout=15)

            if response.status_code == 200:
                data = self._parse_historical_csv(response.text, ticker)

                # Cache the data
                with open(cache_file, 'w') as f:
                    json.dump(data, f)

                return data
        except Exception as e:
            print(f"Historical data error: {e}")

        # Method 2: Generate from daily prices (fallback)
        return self._fetch_daily_prices(ticker, days)

    def get_market_summary(self) -> Dict[str, Any]:
        """Get overall market summary"""
        try:
            url = f"{self.base_url}/market_summary.php"
            response = requests.get(url, timeout=10)

            if response.status_code == 200:
                return self._parse_market_summary(response.text)
        except Exception as e:
            print(f"Market summary error: {e}")

        return {}

    def get_top_stocks(self, category: str = "gainer") -> List[Dict[str, Any]]:
        """Get top gainers/losers/volume"""
        try:
            endpoint_map = {
                "gainer": "top_ten_gainer.php",
                "loser": "top_ten_loser.php",
                "volume": "top_twenty_share.php"
            }

            url = f"{self.base_url}/{endpoint_map.get(category, 'top_ten_gainer.php')}"
            response = requests.get(url, timeout=10)

            if response.status_code == 200:
                return self._parse_top_stocks(response.text)
        except Exception as e:
            print(f"Top stocks error: {e}")

        return []

    def get_company_info(self, ticker: str) -> Dict[str, Any]:
        """Get company fundamentals and info"""
        try:
            url = f"{self.base_url}/displayCompany.php?name={ticker}"
            response = requests.get(url, timeout=10)

            if response.status_code == 200:
                return self._parse_company_info(response.text)
        except Exception as e:
            print(f"Company info error: {e}")

        return {}

    def _parse_dse_current(self, html: str, ticker: str) -> Dict[str, Any]:
        """Parse current price from DSE HTML"""
        # Simple parsing - in production use BeautifulSoup
        import re

        pattern = f"{ticker}.*?([0-9,.]+).*?([0-9,.]+).*?([0-9,.]+)"
        match = re.search(pattern, html)

        if match:
            return {
                "ticker": ticker,
                "last_price": float(match.group(1).replace(',', '')),
                "change": float(match.group(2).replace(',', '')),
                "volume": int(match.group(3).replace(',', ''))
            }
        return {}

    def _parse_price_table(self, html: str, ticker: str) -> Dict[str, Any]:
        """Parse price from HTML table"""
        # Simplified parser - extract price data
        lines = html.split('\n')
        for line in lines:
            if ticker in line:
                # Extract price data from the line
                parts = line.split()
                if len(parts) >= 5:
                    return {
                        "ticker": ticker,
                        "last_price": float(parts[2].replace(',', '')),
                        "high": float(parts[3].replace(',', '')),
                        "low": float(parts[4].replace(',', ''))
                    }
        return {}

    def _parse_historical_csv(self, csv_data: str, ticker: str) -> List[Dict[str, Any]]:
        """Parse historical data from CSV"""
        lines = csv_data.strip().split('\n')
        data = []

        for line in lines[1:]:  # Skip header
            parts = line.split(',')
            if len(parts) >= 6:
                try:
                    data.append({
                        "date": parts[0],
                        "open": float(parts[1]),
                        "high": float(parts[2]),
                        "low": float(parts[3]),
                        "close": float(parts[4]),
                        "volume": int(parts[5])
                    })
                except:
                    continue

        return data

    def _fetch_daily_prices(self, ticker: str, days: int) -> List[Dict[str, Any]]:
        """Fetch daily prices as fallback"""
        data = []
        current_date = datetime.now()

        for i in range(days):
            date = current_date - timedelta(days=i)

            # Skip weekends
            if date.weekday() >= 5:
                continue

            # Try to get price for this date
            price_data = self._get_price_for_date(ticker, date)
            if price_data:
                data.append(price_data)

            # Rate limiting
            time.sleep(0.1)

        return list(reversed(data))

    def _get_price_for_date(self, ticker: str, date: datetime) -> Optional[Dict[str, Any]]:
        """Get price for specific date"""
        # This would fetch from DSE archives
        # For now, return None (would be implemented with actual DSE endpoints)
        return None

    def _parse_market_summary(self, html: str) -> Dict[str, Any]:
        """Parse market summary data"""
        # Extract key metrics
        return {
            "dsex_index": 0,
            "total_volume": 0,
            "total_value": 0,
            "total_trades": 0
        }

    def _parse_top_stocks(self, html: str) -> List[Dict[str, Any]]:
        """Parse top stocks list"""
        stocks = []
        # Parse HTML table for top stocks
        return stocks

    def _parse_company_info(self, html: str) -> Dict[str, Any]:
        """Parse company fundamentals"""
        return {
            "pe_ratio": 0,
            "eps": 0,
            "nav": 0,
            "market_cap": 0
        }


class AlternativeDataFetcher:
    """Fetch data from alternative sources when DSE is unavailable"""

    def __init__(self):
        self.sources = [
            self._fetch_from_yahoo,
            self._fetch_from_investing,
            self._fetch_from_tradingview
        ]

    def get_data(self, ticker: str, exchange: str = "DHA") -> Dict[str, Any]:
        """Try multiple sources to get data"""
        for source in self.sources:
            try:
                data = source(ticker, exchange)
                if data:
                    return data
            except:
                continue
        return {}

    def _fetch_from_yahoo(self, ticker: str, exchange: str) -> Dict[str, Any]:
        """Fetch from Yahoo Finance"""
        yahoo_ticker = f"{ticker}.{exchange}"
        url = f"https://query1.finance.yahoo.com/v8/finance/chart/{yahoo_ticker}"

        response = requests.get(url, headers={'User-Agent': 'Mozilla/5.0'})
        if response.status_code == 200:
            data = response.json()
            return self._parse_yahoo_data(data)
        return {}

    def _fetch_from_investing(self, ticker: str, exchange: str) -> Dict[str, Any]:
        """Fetch from Investing.com"""
        # Would require proper API access
        return {}

    def _fetch_from_tradingview(self, ticker: str, exchange: str) -> Dict[str, Any]:
        """Fetch from TradingView"""
        # Would require API implementation
        return {}

    def _parse_yahoo_data(self, data: dict) -> Dict[str, Any]:
        """Parse Yahoo Finance response"""
        try:
            result = data['chart']['result'][0]
            quotes = result['indicators']['quote'][0]

            return {
                "prices": quotes['close'],
                "volumes": quotes['volume'],
                "timestamps": result['timestamp']
            }
        except:
            return {}


class DataManager:
    """Manage data fetching and storage"""

    def __init__(self, data_dir: str = "data"):
        self.data_dir = Path(data_dir)
        self.data_dir.mkdir(exist_ok=True)

        self.csv_dir = self.data_dir / "csv"
        self.csv_dir.mkdir(exist_ok=True)

        self.scraper = DSEScraper()
        self.alt_fetcher = AlternativeDataFetcher()

    def update_stock_data(self, ticker: str) -> bool:
        """Update data for a stock"""
        print(f"Fetching data for {ticker}...")

        # Get historical data
        historical = self.scraper.get_historical_data(ticker, days=365)

        if not historical:
            # Try alternative sources
            alt_data = self.alt_fetcher.get_data(ticker)
            if alt_data:
                historical = self._convert_alt_data(alt_data)

        if historical:
            # Save to CSV
            df = pd.DataFrame(historical)
            csv_path = self.csv_dir / f"{ticker}_ohlcv.csv"
            df.to_csv(csv_path, index=False)
            print(f"✅ Saved {len(historical)} days of data for {ticker}")
            return True

        print(f"❌ No data found for {ticker}")
        return False

    def update_multiple_stocks(self, tickers: List[str]):
        """Update data for multiple stocks"""
        for ticker in tickers:
            self.update_stock_data(ticker)
            time.sleep(1)  # Rate limiting

    def _convert_alt_data(self, data: dict) -> List[Dict[str, Any]]:
        """Convert alternative data format to standard OHLCV"""
        # Implementation depends on source format
        return []


def main():
    """Main function to fetch DSE data"""

    # Popular DSE stocks
    stocks = [
        "GP",         # Grameenphone
        "SQURPHARMA", # Square Pharma
        "BATBC",      # BAT Bangladesh
        "BRACBANK",   # BRAC Bank
        "CITYBANK",   # City Bank
        "RENATA",     # Renata Limited
        "OLYMPIC",    # Olympic Industries
        "BERGERPBL",  # Berger Paints
        "MARICO",     # Marico Bangladesh
        "ROBI",       # Robi
        "LHBL",       # LafargeHolcim
        "UPGDCL",     # United Power
        "POWERGRID",  # Power Grid
        "BSCCL",      # Bangladesh Submarine Cable
        "EBL",        # Eastern Bank
    ]

    manager = DataManager()

    print("=" * 50)
    print("DSE Data Scraper - Fetching Stock Data")
    print("=" * 50)

    manager.update_multiple_stocks(stocks)

    print("\n✅ Data fetching complete!")
    print(f"📁 Data saved in: {manager.csv_dir}")
    print("\n🚀 Restart Docker to use the new data:")
    print("   docker-compose restart")


if __name__ == "__main__":
    main()

File: packages/scraper/src/dse-scraper.ts
import { join } from 'node:path';
import {
  ensureDir,
  fetchText,
  formatDate,
  parseHistoricalCsv,
  readJsonCache,
  sleep,
  writeJsonCache,
  type OhlcvRow,
} from './utils.js';

export interface LatestPrice {
  ticker: string;
  last_price?: number;
  change?: number;
  volume?: number;
  high?: number;
  low?: number;
}

export class DSEScraper {
  readonly cacheDir: string;
  readonly baseUrl = 'https://www.dsebd.org';

  constructor(cacheDir = 'data/cache') {
    this.cacheDir = cacheDir;
    ensureDir(cacheDir);
  }

  async getLatestPrice(ticker: string): Promise<Record<string, unknown>> {
    try {
      const url = `${this.baseUrl}/php_graph/market_price.php`;
      const data = await fetchText(url);
      if (data) {
        return this.parseDseCurrent(data, ticker);
      }
    } catch (e) {
      console.error('DSE API error:', e);
    }

    try {
      const url = `${this.baseUrl}/latest_share_price_scroll_l.php`;
      const data = await fetchText(url);
      if (data) {
        return this.parsePriceTable(data, ticker);
      }
    } catch (e) {
      console.error('Alternative endpoint error:', e);
    }

    return {};
  }

  async getHistoricalData(ticker: string, days = 365): Promise<OhlcvRow[]> {
    const cacheFile = join(this.cacheDir, `${ticker}_history.json`);
    const cached = readJsonCache<OhlcvRow[]>(cacheFile, 3600);
    if (cached) {
      return cached;
    }

    try {
      const endDate = new Date();
      const startDate = new Date(endDate.getTime() - days * 24 * 60 * 60 * 1000);

      const params = new URLSearchParams({
        startDate: formatDate(startDate),
        endDate: formatDate(endDate),
        inst: ticker,
        archive: 'data',
      });

      const url = `${this.baseUrl}/day_end_archive.php?${params.toString()}`;
      const responseText = await fetchText(url);

      if (responseText) {
        const data = this.parseHistoricalCsv(responseText, ticker);
        writeJsonCache(cacheFile, data);
        return data;
      }
    } catch (e) {
      console.error('Historical data error:', e);
    }

    return this.fetchDailyPrices(ticker, days);
  }

  async getMarketSummary(): Promise<Record<string, unknown>> {
    try {
      const url = `${this.baseUrl}/market_summary.php`;
      const html = await fetchText(url);
      if (html) {
        return this.parseMarketSummary(html);
      }
    } catch (e) {
      console.error('Market summary error:', e);
    }
    return {};
  }

  async getTopStocks(category = 'gainer'): Promise<Record<string, unknown>[]> {
    try {
      const endpointMap: Record<string, string> = {
        gainer: 'top_ten_gainer.php',
        loser: 'top_ten_loser.php',
        volume: 'top_twenty_share.php',
      };

      const endpoint = endpointMap[category] ?? 'top_ten_gainer.php';
      const url = `${this.baseUrl}/${endpoint}`;
      const html = await fetchText(url);
      if (html) {
        return this.parseTopStocks(html);
      }
    } catch (e) {
      console.error('Top stocks error:', e);
    }
    return [];
  }

  async getCompanyInfo(ticker: string): Promise<Record<string, unknown>> {
    try {
      const url = `${this.baseUrl}/displayCompany.php?name=${ticker}`;
      const html = await fetchText(url);
      if (html) {
        return this.parseCompanyInfo(html);
      }
    } catch (e) {
      console.error('Company info error:', e);
    }
    return {};
  }

  parseDseCurrent(html: string, ticker: string): Record<string, unknown> {
    const pattern = new RegExp(`${ticker}.*?([0-9,.]+).*?([0-9,.]+).*?([0-9,.]+)`);
    const match = html.match(pattern);

    if (match) {
      return {
        ticker,
        last_price: parseFloat(match[1]!.replace(/,/g, '')),
        change: parseFloat(match[2]!.replace(/,/g, '')),
        volume: parseInt(match[3]!.replace(/,/g, ''), 10),
      };
    }
    return {};
  }

  parsePriceTable(html: string, ticker: string): Record<string, unknown> {
    const lines = html.split('\n');
    for (const line of lines) {
      if (line.includes(ticker)) {
        const parts = line.split(/\s+/);
        if (parts.length >= 5) {
          return {
            ticker,
            last_price: parseFloat(parts[2]!.replace(/,/g, '')),
            high: parseFloat(parts[3]!.replace(/,/g, '')),
            low: parseFloat(parts[4]!.replace(/,/g, '')),
          };
        }
      }
    }
    return {};
  }

  parseHistoricalCsv(csvData: string, _ticker: string): OhlcvRow[] {
    return parseHistoricalCsv(csvData);
  }

  async fetchDailyPrices(ticker: string, days: number): Promise<OhlcvRow[]> {
    const data: OhlcvRow[] = [];
    const currentDate = new Date();

    for (let i = 0; i < days; i++) {
      const date = new Date(currentDate.getTime() - i * 24 * 60 * 60 * 1000);

      if (date.getDay() >= 5) {
        continue;
      }

      const priceData = await this.getPriceForDate(ticker, date);
      if (priceData) {
        data.push(priceData);
      }

      await sleep(100);
    }

    return data.reverse();
  }

  async getPriceForDate(_ticker: string, _date: Date): Promise<OhlcvRow | null> {
    return null;
  }

  parseMarketSummary(_html: string): Record<string, unknown> {
    return {
      dsex_index: 0,
      total_volume: 0,
      total_value: 0,
      total_trades: 0,
    };
  }

  parseTopStocks(_html: string): Record<string, unknown>[] {
    return [];
  }

  parseCompanyInfo(_html: string): Record<string, unknown> {
    return {
      pe_ratio: 0,
      eps: 0,
      nav: 0,
      market_cap: 0,
    };
  }
}

export class AlternativeDataFetcher {
  async getData(ticker: string, exchange = 'DHA'): Promise<Record<string, unknown>> {
    const sources = [
      () => this.fetchFromYahoo(ticker, exchange),
      () => this.fetchFromInvesting(ticker, exchange),
      () => this.fetchFromTradingview(ticker, exchange),
    ];

    for (const source of sources) {
      try {
        const data = await source();
        if (data && Object.keys(data).length > 0) {
          return data;
        }
      } catch {
        continue;
      }
    }
    return {};
  }

  async fetchFromYahoo(ticker: string, exchange: string): Promise<Record<string, unknown>> {
    const yahooTicker = `${ticker}.${exchange}`;
    const url = `https://query1.finance.yahoo.com/v8/finance/chart/${yahooTicker}`;

    const response = await fetch(url, { headers: { 'User-Agent': 'Mozilla/5.0' } });
    if (response.ok) {
      const data = (await response.json()) as Record<string, unknown>;
      return this.parseYahooData(data);
    }
    return {};
  }

  async fetchFromInvesting(_ticker: string, _exchange: string): Promise<Record<string, unknown>> {
    return {};
  }

  async fetchFromTradingview(_ticker: string, _exchange: string): Promise<Record<string, unknown>> {
    return {};
  }

  parseYahooData(data: Record<string, unknown>): Record<string, unknown> {
    try {
      const chart = data.chart as Record<string, unknown>;
      const result = (chart.result as Record<string, unknown>[])[0]!;
      const quotes = (result.indicators as Record<string, unknown>).quote as Record<string, unknown>[];
      const quote = quotes[0]!;

      return {
        prices: quote.close,
        volumes: quote.volume,
        timestamps: result.timestamp,
      };
    } catch {
      return {};
    }
  }
}

export class DataManager {
  readonly dataDir: string;
  readonly csvDir: string;
  readonly scraper: DSEScraper;
  readonly altFetcher: AlternativeDataFetcher;

  constructor(dataDir = 'data') {
    this.dataDir = dataDir;
    this.csvDir = join(dataDir, 'csv');
    ensureDir(this.dataDir);
    ensureDir(this.csvDir);
    this.scraper = new DSEScraper();
    this.altFetcher = new AlternativeDataFetcher();
  }

  async updateStockData(ticker: string): Promise<boolean> {
    console.log(`Fetching data for ${ticker}...`);

    let historical = await this.scraper.getHistoricalData(ticker, 365);

    if (historical.length === 0) {
      const altData = await this.altFetcher.getData(ticker);
      if (altData && Object.keys(altData).length > 0) {
        historical = this.convertAltData(altData);
      }
    }

    if (historical.length > 0) {
      const { writeCsv } = await import('./utils.js');
      const csvPath = join(this.csvDir, `${ticker}_ohlcv.csv`);
      writeCsv(csvPath, historical, ['date', 'open', 'high', 'low', 'close', 'volume']);
      console.log(`✅ Saved ${historical.length} days of data for ${ticker}`);
      return true;
    }

    console.log(`❌ No data found for ${ticker}`);
    return false;
  }

  async updateMultipleStocks(tickers: string[]): Promise<void> {
    for (const ticker of tickers) {
      await this.updateStockData(ticker);
      await sleep(1000);
    }
  }

  convertAltData(_data: Record<string, unknown>): OhlcvRow[] {
    return [];
  }
}

export const POPULAR_DSE_STOCKS = [
  'GP',
  'SQURPHARMA',
  'BATBC',
  'BRACBANK',
  'CITYBANK',
  'RENATA',
  'OLYMPIC',
  'BERGERPBL',
  'MARICO',
  'ROBI',
  'LHBL',
  'UPGDCL',
  'POWERGRID',
  'BSCCL',
  'EBL',
] as const;

export async function runDseScraperMain(): Promise<void> {
  const manager = new DataManager();

  console.log('='.repeat(50));
  console.log('DSE Data Scraper - Fetching Stock Data');
  console.log('='.repeat(50));

  await manager.updateMultipleStocks([...POPULAR_DSE_STOCKS]);

  console.log('\n✅ Data fetching complete!');
  console.log(`📁 Data saved in: ${manager.csvDir}`);
  console.log('\n🚀 Restart Docker to use the new data:');
  console.log('   docker-compose restart');
}
